src/openai/_tracing.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import json
+import functools
+from typing import TypeVar, Callable, Any, Generator, Iterator, Union, Literal
+from typing_extensions import ParamSpec
+
+from ._extras import opentelemetry as trace, has_tracing_enabled
+from .types.chat import ChatCompletion, ChatCompletionChunk
+from .types.completion import Completion
+from ._streaming import Stream
+
+TracedMessageTypes = Literal["system", "user", "assistant", "tool"]
+TracedModels = Union[ChatCompletion, Completion]
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+def _set_attribute(span: trace.Span, key: str, value: Any) -> None:
+    if value:
+        span.set_attribute(key, value)
+
+
+def _add_request_chat_message_event(span: trace.Span, **kwargs: Any) -> None:
+    for message in kwargs.get("messages", []):
+        try:
+            message = message.to_dict()
+        except AttributeError:
+            pass
+        
+        if message.get("role") in TracedMessageTypes:
```

Comment:
I would probably add the span attribute even if the role is not a known one. If there are ever additional roles added, we shouldn't have to update this code.